### ISSUE TITLE
dependance on chemaxon is annoying

### ISSUE THREAD

**Issue by UnixJunkie** [ci-lab-cz/easydock#17]
maybe switch to Dimorphite-DL:
https://jcheminf.biomedcentral.com/articles/10.1186/s13321-019-0336-9

I am in academia and I don't even have a chemaxon license anymore...
Software vendors always change their license terms one day or another...

**Comment by DrrDom** [comment]
Agree, chemaxon change of an academic license is inappropriate. I now consider different options. Previously I tested Dimophite-DL, pkasolver and chemaxon and at the first glance chemaxon performed better in complex cases. However, I would prefer a more systematic study. At least it would be reasonable to compare Dimorphite-DL and pkasolver. Unfortunately, this will take time, so it will not be fast, but we will try different alternatives. Actually, this could be an opportunity to develop something new

**Comment by DrrDom** [comment]
It is still possible to disable chemaxon protonation and use some external tools to protonate structures before docking

**Comment by UnixJunkie** [comment]
as another hackish alternative, there would be obabel's -p option (Add Hydrogens appropriate for pH)

**Comment by DrrDom** [comment]
I implemented an easy way to integrate protonation tools. I also implemented dimorphite as an example in the separate branch `dimorphite-example`.

What I dislike with dimorphite
- it is not a package and integration of third-party programs as an integral part make it hard to update and support. Dimorphite is pretty old and currently is not developed, but avoiding inclusion of third-party software is a general rule.
- I did not test, but dimorphite may be slow on large sets of compounds, because it works in a single process.
- license of dimorphite is Apache 2 and differs from easydock BSD-3. They are compatible, but if it will become an integral part, easydock should have two licenses.
- it is impossible to use dimorphite API (function `run`), because it catches argparse of main program arguments. This requires fixes of dimorphite code.

I would prefer if dimorphite will be converted into a package to be installed by pip and can be run using `pool` on multiple cores. Ideally, if a bug with catching the main argparser using dimorphite API will be also fixed.

If someone will help with that, it will be faster to integrate dimorphote.

@Feriolet, would you be able to help with that?

**Comment by Feriolet** [comment]
Yes, these have been some of the pain points of using Dimorphite-DL for me as well. I have tried to run 250k compounds, but the `init_db` ran slowly (around 7 hours). Dimirphite-DL may be the cause, but I have not tested it.

It is also annoying the argparse of Dimorphite is problematic so I cant be used it as a package, and the current solution I have uses subprocess.

I can try to fix the bug of Dimorphite_DL

**Comment by Feriolet** [comment]
I have found the reason for the argparse error for dimorphite_dl. The main reason that the dimorphite catches the main program argument is because of the command line `args = vars(parser.parse_args())` where `.parse_args()` collect the argument from the terminal as the default (main run_dock argument). To override the default setting and use the desired argument, we have to put the list `['--argument',value,'--argument',value]` inside the `.parse_args(list)`.

Since the `**kwargs` returns a dict_output, I have made a temporary function that converts the dict into a flat list:

```python
def convert_kwargs_to_sys_argv(params: dict) -> list:
    sys_argv_list = []
    for key, value in params.items():
        # Since some argument have store_true argument, we don't need to include the 'True' value into the sys.argv
        if value is True:
            sys_argv_list.append('--'+key)
        elif value is False:
            continue
        else:
            sys_argv_list.append('--'+key)
            sys_argv_list.append(str(value))
        
    return sys_argv_list
```

Then, the `main()` function of the dimorphite_dl is modified to:
```python
def main(params=None):
    """The main definition run when you call the script from the commandline.

    :param params: The parameters to use. Entirely optional. If absent,
                   defaults to None, in which case argments will be taken from
                   those given at the command line.
    :param params: dict, optional
    :return: Returns a list of the SMILES strings return_as_list parameter is
             True. Otherwise, returns None.
    """
    sys_argv_list = convert_kwargs_to_sys_argv(params)
    parser = ArgParseFuncs.get_args()
    args = vars(parser.parse_args(sys_argv_list))
    print(args)
    if not args["silent"]:
        print_header()

    # Add in any parameters in params.
    if params is not None:
        for k, v in params.items():
            args[k] = v

    # If being run from the command line, print out all parameters.
    if __name__ == "__main__":
        if not args["silent"]:
            print("\nPARAMETERS:\n")
            for k in sorted(args.keys()):
                print(k.rjust(13) + ": " + str(args[k]))
            print("")

    if args["test"]:
        # Run tests.
        TestFuncs.test()
    else:
        # Run protonation
        if "output_file" in args and args["output_file"] is not None:
            # An output file was specified, so write to that.
            with open(args["output_file"], "w") as file:
                for protonated_smi in Protonate(args):
                    file.write(protonated_smi + "\n")
        elif "return_as_list" in args and args["return_as_list"] == True:
            return list(Protonate(args))
        else:
            # No output file specified. Just print it to the screen.
            for protonated_smi in Protonate(args):
                print(protonated_smi)
```

**Comment by DrrDom** [comment]
Good first step. Thank you!

I checked the code of Dimorphite for parallelization and may say it can be not the most difficult part. The code of a generator class looks quite complex and I do not understand whether this complexity is necessary or not.

**Comment by Feriolet** [comment]
Yes, it will take time for me to understand the code as it uses multiple Classes in huge chunk as well.

**Comment by Feriolet** [comment]
I noticed that the `Protonate()` class is an iterator with the `__next()__` and `__iter()__`. Is it even possible to use multiprocessing for this type of class? Calling the `Protonate(args)` may immediately loop the smi list before we can multiprocess it.

**Comment by DrrDom** [comment]
In theory it should be possible, but I do not have experience with such classes. I usually implement such kind of things in a different way

A very simple example:

```python
with Pool(ncpu) as pool:
    with open(smi_filename) as f:
        for mol_id, protonated_smi in pool.imap_unordered(Protonate, f):
             if protonated_smi:
                  update_db(mol_id, protonated_smi)
```

where `Protonate` would take a line and returns `mol_id` and `protonated_smi`. `f` is actually a generator that avoids memory overflow.

There are some examples of generators working with multiprocessing - https://codereview.stackexchange.com/questions/259029/compute-the-outputs-of-a-generator-in-parallel, but those implementations are also simpler.

**Comment by Feriolet** [comment]
Yep, I guess it is possible. I made a simple script with a random Iterator class to test the multiprocess, and it seems that it works

```python
from multiprocessing import Pool
import multiprocessing
import time

class Test:
 
    # Constructor
    def __init__(self, limit):
        self.limit = limit
 
    # Creates iterator object
    # Called when iteration is initialized
    def __iter__(self):
        self.x = 10
        return self
 
    # To move to next element. In Python 3,
    # we should replace next with __next__
    def __next__(self):
        time.sleep(5) #Assuming the function take 5 seconds
        # Store current value ofx
        x = self.x
 
        # Stop iteration if limit is reached
        if x > self.limit:
            raise StopIteration
 
        # Else increment and return old value
        self.x = x + 1;
        return x
    
    def run(self):
        with Pool(3) as pool:
            return pool.imap_unordered(self.__next__, self)
    


if __name__ == '__main__':
if __name__ == '__main__':
    start1 = time.time()
    for x in Test(30):
        pass
    end1 = time.time()
    print(end1-start1)

    start2 = time.time()
    test = Test(30)
    result = test.run()
    end2 = time.time()

    print('hello')
    print(end2 - start2)
```
```txt
110.08608412742615
hello
5.082077980041504
```

**Comment by Feriolet** [comment]
I have tried to understand how iterator class work and I think I got a few ideas on what to do. The dimorphite class needs to take the input filename, and the **iterator** is generated by the `LoadSMIFile()` class and stored in the `args["smiles_and_data]` when the `ArgParseFuncs.clean_args()` is called. Because of the nature of Iterator class, the input must be an iterator type, not iterable type.

The above code probably won't work because the `Protonate()` will treat `f` as the `args` input, which is not what we wanted.

My idea is that we can either change how the `Protonate()` treats the input by changing the `args["smiles_and_data"]` code, but it may break how dimorphite intends to run the code. Another ridiculous way is that we can split the smi files to `n` number of smi (e.g., 250k smi will have 250k separate files), which can be easily multiprocess and we don't need to modify the code too much. But again, generating 250k files is kind of ridiculous.

**Comment by DrrDom** [comment]
The idea of multiple input files is very nice, because we nevertheless will store all these data on a disk and it does not matter in how many files.

We can create 2x-3x files as the number of CPUs submitted by a user in the command line (argument `-c`), propagate this `ncpu` argument to the `add_protonation` function and use it to run multiples input files using `pool` and `subprocess`. Not a perfect solution but fast and working.

If the speed will be too slow, we may think to use `dask` for this task, but I hope it will be reasonably good.

**Comment by DrrDom** [comment]
You may create a package of dimorphite-dl under your account. Uploading it to the pypi is not necessary, we may specify in README to install dimorphite directly from your repository.

**Comment by DrrDom** [comment]
I mean `pip install git+...`. So I suggest to make it similarly to `easydock`. Then we will be able to import it from `easydock`and run directly from Python even without `subprocess` (since we will solve the issue with command line arguments parsing).

**Comment by Samuel-gwb** [comment]
Is Dimorphite ready to use in the branch ?

**Comment by Feriolet** [comment]
> I mean `pip install git+...`. So I suggest to make it similarly to `easydock`. Then we will be able to import it from `easydock`and run directly from Python even without `subprocess` (since we will solve the issue with command line arguments parsing).

Oh okay, but how is it different than using the dimorphite_example currently? I am forking that branch to modify the code to implement the current feature. I will try to make a pull request soon to fix the dimorphite implementation

**Comment by DrrDom** [comment]
@Samuel-gwb, yes, but it uses one core and for large sets this will take some time. 250k molecules ~ 20 min.

@Feriolet, the difference will be that we can completely remove directory and files of dimorphite from easydock package. Dimorphite will be installed as an individual package and imported similarly to the current import (only the name of the package will change) If you need more details, please ask

**Comment by Feriolet** [comment]
But we do still need to change the dimorphite source code right? I feel that it is inconvenient for the user to install it as a separate package and then alter it by themselves. Or is there something that I missed about installing it in a different directory from easydock?

**Comment by DrrDom** [comment]
Yes, the source code should be modified to enable mutiprocessing and correct treatment of command line arguments. A separate package will not solve the issue with command line arguments, so this should be explicitly fixed.

Installation as a package is preferable and convenient.
1. No license mixture needed.
2. Maintain a separate package is simpler. It may also be further developed independently, if someone will find these changes useful.
3. A user nevertheless installs a number of packages. This will require to copy-paste one line from README.

**Comment by Feriolet** [comment]
I can't install dimorphite through pip.

`ERROR: git+https://github.com/durrantlab/dimorphite_dl does not appear to be a Python project: neither 'setup.py' nor 'pyproject.toml' found.`

**Comment by DrrDom** [comment]
Yes, of course. Therefore, it should be converted into a Python package to enable installation from pip.

This is quite easy. You may check the structure of `easydock` repository. You have to add `setup.cfg`, `setup.py`, `LICENSE.txt`, `README.md`, `MANIFEST.md` to the root dir of your fork of the repository, move other files in a subdirectory named for example `dimorphite` and add in that subdirectory file `__init_.py` (it may be empty).

In the `MANIFEST.md` you should list files which are not python scripts (*.py), but which should be supplied with the package. In dimorpfite this will be the file with patterns (at least for the first glance).
Example if the file - https://github.com/DrrDom/pmapper/blob/master/MANIFEST.in

**Comment by Feriolet** [comment]
Huh I tried creating the `setup.cfg` and `setup.py` for the dimorphite but it does not create a library in the conda environment (unless you considered `egg-link` to be one). I made the `dimorphite` directory in my `Desktop` directory, then install it with `pip install -e .`

It works for now though. I have tried re-running it and it works. I forgot what min and max pH I used for my initial run, so I can't test if my code in my pull request actually affect the docking score. As I can't reproduce the same protonated smi and docking score for my code, maybe you can see if the code I send in the pull request is good enough.

Also, since we are creating the `setup.cfg` and `setup.py`, are we going to include them into our repository as well, or do we ask the user to do it on their own in the `README` file?

**Comment by DrrDom** [comment]
Installing with `pip install -e .` does not transfer files to the conda environment, but create links. This is almost like an ordinary installation. So, everything is OK.

You have to create your fork on github with `dimorhite` code and edit it to fix problems which interfere with `easydock`. Afterwards you can make PR to the original repository, If maintainers will accept it this will be excellent. If not, we can mention in `easydock` `README` to install `dimorphite` from `pip git+https://github.com/Feriolet/dimorphite_dl`. In the latter case you will become the maintainer of this specific version of dimorphite.
So, no `dimorhite` files should be included in `easydock` repo.

**Comment by Samuel-gwb** [comment]
Seems that successfully using dimorphite_dl by "pip git+https://github.com/Feriolet/dimorphite_dl".
Great work!
Update the official code according to the branches?

**Comment by DrrDom** [comment]
Yes, tests were passed successfully. I merged all changes to master, but I'll postpone to release a new version, because there is one critical issue which should be somehow addressed.

i noticed that Dimorphite does not behave well with nitrogen centers and molecules having more than a single protonation center. Below are several test examples. In the cases 3 and 4 the nitrogen atom was not protonated as expected.
```
smi			smi_protonated
Oc1ccccn1		[O-]c1ccccn1
O=c1cccc[nH]1		O=c1cccc[n-]1
NCC(=O)O		NCC(=O)[O-]
C[C@@H]1CCCN(C)C1	C[C@@H]1CCCN(C)C1
```
Dimorphite was designed to enumerate a list of possible forms at a given pH and it cannot predict the most favorable one. The logic behind was to enumerate all reasonable forms and dock all of them. Since we limit the number of protonated forms to 1, Dimorphite takes the first one from the list and returns it.

Therefore, we have two options:
- use dimorphite as is (and probably get incorrect protonation states for some random molecules) -  this looks very misleading and unexpected to a user
- re-implement treating of protonation forms (similarly as stereoisomers) to consider all of them - this will increase complexity of manipulation with database records. Docking of all possible protonation forms may be also misleading in my opinion, but probably less than docking of a single incorrect form.

Alternative solution - pkasolver.
Half a year ago I adapted pkasolver to predict the most stable protonation state. I took a form with pKa closest to 7.4. pkasolver also has some drawbacks and errors in assignment of protonation states. However, to some extent it can treat molecules with multiple centers (but the model which is publicly available was trained on single-center molecules only). I have to adapt previous implementation to the new program architecture and can make a PR to enable discussion of all solutions.

Therefore, in the beginning I mentioned that it would be good to study and compare all available solutions (free or open-source) to choose more appropriate one(s).

**Comment by DrrDom** [comment]
Yes, this is pkasolver repo.
It takes a single molecule and enumerates multiple protonation forms with predicted pKa and I chose the one closest to 7.4.
I'll try to adapt it quickly and make PR to show the idea and test it. With the new architecture it should be relatively easy to do.

You are right, I missed the issue of too many forms for a single molecule. It may really substantially increase docking time and may not be necessity. This is another cons.

**Comment by DrrDom** [comment]
I uploaded the branch `pkasolver2`. It is expected to work, however, I could not install dependencies for `pkasolver`, therefore I did not test it, but previously almost the same code worked.
You may play with it. I remember also some strange behavior.
To run it just specify `run_dock ... --protonation pkasolver`

**Comment by Samuel-gwb** [comment]
Got it! thanks

**Comment by Feriolet** [comment]
In case you are testing the code very soon, do change the code of `protonation.py` line 11 from `from read_input import read_input` to `from easydock.read_input import read_input` to prevent the import error

**Comment by DrrDom** [comment]
Thank you, fixed!
If someone will test it, please report whether it works and how.

I cannot create an environment from the yml-file and manual installation requires specific versions of packages because on my old system not all recent versions can be installed.

**Comment by Samuel-gwb** [comment]
After download pkasolver2, 
conda activate easydock_pka 
-- install dependencies --
1) pip install torch==1.13.1+cu116 torchvision==0.14.1+cu116 torchaudio==0.13.1 --extra-index-url https://download.pytorch.org/whl/cu116
2) pip install torch-geometric==2.0.1 torch-sparse torch-scatter molvs chembl_webresource_client
3) conda install -c conda-forge matplotlib pytest-cov codecov svgutils cairosvg ipython
4) python setup.py install

Then run_dock ... --protonation pkasolver -c 1 --sdf

Errors appear:
###############################################
Process ForkPoolWorker-1:
Traceback (most recent call last):
  File "/home/gwb/miniconda3/envs/easydock_pka/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/home/gwb/miniconda3/envs/easydock_pka/lib/python3.9/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/gwb/miniconda3/envs/easydock_pka/lib/python3.9/multiprocessing/pool.py", line 114, in worker
    task = get()
  File "/home/gwb/miniconda3/envs/easydock_pka/lib/python3.9/multiprocessing/queues.py", line 367, in get
    return _ForkingPickler.loads(res)
  File "/home/gwb/miniconda3/envs/easydock_pka/lib/python3.9/site-packages/torch/multiprocessing/reductions.py", line 120, in rebuild_cuda_tensor
    torch.cuda._lazy_init()
  File "/home/gwb/miniconda3/envs/easydock_pka/lib/python3.9/site-packages/torch/cuda/__init__.py", line 217, in _lazy_init
    raise RuntimeError(
RuntimeError: Cannot re-initialize CUDA in forked subprocess. To use CUDA with multiprocessing, you must use the 'spawn' start method
###############################################

Code in __init__.py is as (raise RuntimeError ... ):
###############################################
def _lazy_init():
    global _initialized, _queued_calls
    if is_initialized() or hasattr(_tls, 'is_initializing'):
        return
    with _initialization_lock:
        # We be double-checked locking, boys!  This is OK because
        # the above test was GIL protected anyway.  The inner test
        # is for when a thread blocked on some other thread which was
        # doing the initialization; when they get the lock, they will
        # find there is nothing left to do.
        if is_initialized():
            return
        # It is important to prevent other threads from entering _lazy_init
        # immediately, while we are still guaranteed to have the GIL, because some
        # of the C calls we make below will release the GIL
        if _is_in_bad_fork():
            raise RuntimeError(
                "Cannot re-initialize CUDA in forked subprocess. To use CUDA with "
                "multiprocessing, you must use the 'spawn' start method")
###############################################

**Comment by Samuel-gwb** [comment]
Is it possible to carry out pka_solver before run_dock?
Something like: 
ensemble.smi  --> ensemble_pka.smi
and then  run_dock ensemble_pka.smi ...

And
ensemble.smi  --> ensemble_pka.smi
means, due to pka_solver :
example_1           --> example_1 
example_2           --> example_2_01
                                   example_2_02
                                   ...
...

**Comment by Feriolet** [comment]
Can I know how many GPU do you have on your machine?

It is possible to protonate your smiles first with separate script and then run the `run_dock` with the `--no_protonation` argument

**Comment by Samuel-gwb** [comment]
2 GPUs. Need specification?

**Comment by UnixJunkie** [comment]
Depending on a GPU might also be annoying.
It would be nice if the code can run on a regular laptop, which might not
have any decent GPU.
Some cluster nodes have very powerful CPUs but no GPU, etc.

**Comment by Feriolet** [comment]
yes that is true. I have not install the pkasolver because of CUDA compatibility issue right now, but if I may know, does pkasolver use cpu or gpu to run the program? If possible, we can force the pkasolver to use cpu instead of gpu to make our code consistent

**Comment by Samuel-gwb** [comment]
Yes. By default, pkasolver use gpu, indicated by nvidia-smi.

**Comment by Feriolet** [comment]
@Samuel-gwb can you provide me with the full torch version and CUDA? I tried to use your version mentioned above but I get a `segmentation fault (core dumped)` error

**Comment by Feriolet** [comment]
The `QueryModel()` class seems to accomodate both GPU and CPU as can be seen in this code:

```python
            if torch.cuda.is_available() == False:  # If only CPU is available
                checkpoint = torch.load(
                    f"{base_path}/trained_model_without_epik/best_model_{i}.pt",
                    map_location=torch.device("cpu"),
                )
            else:
                checkpoint = torch.load(
                    f"{base_path}/trained_model_without_epik/best_model_{i}.pt"
                )
```

**Comment by Samuel-gwb** [comment]
Please check the attachment

[environment_easydock_pka.txt](https://github.com/ci-lab-cz/easydock/files/14913718/environment_easydock_pka.txt)

**Comment by Samuel-gwb** [comment]
It seems that running on cpu has problem. I created a new environment with cpu version torch. 
However, errors appear as, using previous run_dock ... --protonation pkasolver :

  File "/home/gwb/miniconda3/envs/easydock_pkaCPU/lib/python3.9/site-packages/torch/nn/modules/module.py", line 2153, in load_state_dict
    raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
RuntimeError: Error(s) in loading state_dict for GINPairV1:
        Missing key(s) in state_dict: "GIN_p.convs.0.nn.lins.0.weight", "GIN_p.convs.0.nn.lins.0.bias", "GIN_p.convs.0.nn.lins.1.weight", "GIN_p.convs.0.nn.lins.1.bias", "GIN_p.convs.1.nn.lins.0.weight", "GIN_p.convs.1.nn.lins.0.bias", "GIN_p.convs.1.nn.lins.1.weight", "GIN_p.convs.1.nn.lins.1.bias", "GIN_p.convs.2.nn.lins.0.weight", "GIN_p.convs.2.nn.lins.0.bias", "GIN_p.convs.2.nn.lins.1.weight", "GIN_p.convs.2.nn.lins.1.bias", "GIN_p.convs.3.nn.lins.0.weight", "GIN_p.convs.3.nn.lins.0.bias", "GIN_p.convs.3.nn.lins.1.weight", "GIN_p.convs.3.nn.lins.1.bias", "GIN_d.convs.0.nn.lins.0.weight", "GIN_d.convs.0.nn.lins.0.bias", "GIN_d.convs.0.nn.lins.1.weight", "GIN_d.convs.0.nn.lins.1.bias", "GIN_d.convs.1.nn.lins.0.weight", "GIN_d.convs.1.nn.lins.0.bias", "GIN_d.convs.1.nn.lins.1.weight", "GIN_d.convs.1.nn.lins.1.bias", "GIN_d.convs.2.nn.lins.0.weight", "GIN_d.convs.2.nn.lins.0.bias", "GIN_d.convs.2.nn.lins.1.weight", "GIN_d.convs.2.nn.lins.1.bias", "GIN_d.convs.3.nn.lins.0.weight", "GIN_d.convs.3.nn.lins.0.bias", "GIN_d.convs.3.nn.lins.1.weight", "GIN_d.convs.3.nn.lins.1.bias".
        Unexpected key(s) in state_dict: "GIN_p.lin.weight", "GIN_p.lin.bias", "GIN_p.convs.0.nn.0.weight", "GIN_p.convs.0.nn.0.bias", "GIN_p.convs.0.nn.1.weight", "GIN_p.convs.0.nn.1.bias", "GIN_p.convs.0.nn.1.running_mean", "GIN_p.convs.0.nn.1.running_var", "GIN_p.convs.0.nn.1.num_batches_tracked", "GIN_p.convs.0.nn.3.weight", "GIN_p.convs.0.nn.3.bias", "GIN_p.convs.1.nn.0.weight", "GIN_p.convs.1.nn.0.bias", "GIN_p.convs.1.nn.1.weight", "GIN_p.convs.1.nn.1.bias", "GIN_p.convs.1.nn.1.running_mean", "GIN_p.convs.1.nn.1.running_var", "GIN_p.convs.1.nn.1.num_batches_tracked", "GIN_p.convs.1.nn.3.weight", "GIN_p.convs.1.nn.3.bias", "GIN_p.convs.2.nn.0.weight", "GIN_p.convs.2.nn.0.bias", "GIN_p.convs.2.nn.1.weight", "GIN_p.convs.2.nn.1.bias", "GIN_p.convs.2.nn.1.running_mean", "GIN_p.convs.2.nn.1.running_var", "GIN_p.convs.2.nn.1.num_batches_tracked", "GIN_p.convs.2.nn.3.weight", "GIN_p.convs.2.nn.3.bias", "GIN_p.convs.3.nn.0.weight", "GIN_p.convs.3.nn.0.bias", "GIN_p.convs.3.nn.1.weight", "GIN_p.convs.3.nn.1.bias", "GIN_p.convs.3.nn.1.running_mean", "GIN_p.convs.3.nn.1.running_var", "GIN_p.convs.3.nn.1.num_batches_tracked", "GIN_p.convs.3.nn.3.weight", "GIN_p.convs.3.nn.3.bias", "GIN_d.lin.weight", "GIN_d.lin.bias", "GIN_d.convs.0.nn.0.weight", "GIN_d.convs.0.nn.0.bias", "GIN_d.convs.0.nn.1.weight", "GIN_d.convs.0.nn.1.bias", "GIN_d.convs.0.nn.1.running_mean", "GIN_d.convs.0.nn.1.running_var", "GIN_d.convs.0.nn.1.num_batches_tracked", "GIN_d.convs.0.nn.3.weight", "GIN_d.convs.0.nn.3.bias", "GIN_d.convs.1.nn.0.weight", "GIN_d.convs.1.nn.0.bias", "GIN_d.convs.1.nn.1.weight", "GIN_d.convs.1.nn.1.bias", "GIN_d.convs.1.nn.1.running_mean", "GIN_d.convs.1.nn.1.running_var", "GIN_d.convs.1.nn.1.num_batches_tracked", "GIN_d.convs.1.nn.3.weight", "GIN_d.convs.1.nn.3.bias", "GIN_d.convs.2.nn.0.weight", "GIN_d.convs.2.nn.0.bias", "GIN_d.convs.2.nn.1.weight", "GIN_d.convs.2.nn.1.bias", "GIN_d.convs.2.nn.1.running_mean", "GIN_d.convs.2.nn.1.running_var", "GIN_d.convs.2.nn.1.num_batches_tracked", "GIN_d.convs.2.nn.3.weight", "GIN_d.convs.2.nn.3.bias", "GIN_d.convs.3.nn.0.weight", "GIN_d.convs.3.nn.0.bias", "GIN_d.convs.3.nn.1.weight", "GIN_d.convs.3.nn.1.bias", "GIN_d.convs.3.nn.1.running_mean", "GIN_d.convs.3.nn.1.running_var", "GIN_d.convs.3.nn.1.num_batches_tracked", "GIN_d.convs.3.nn.3.weight", "GIN_d.convs.3.nn.3.bias".

**Comment by Feriolet** [comment]
It seems that this will be quite troublesome.

I have finally solved the seg fault error by using a different version of torch by running the following code:
`conda install pytorch==1.11.0 pytorch-cuda=11.7 pytorch_scatter pytorch_sparse pytorch_cluster pytorch_spline_conv pyg -c pytorch -c nvidia -c pyg`

The `RuntimeError: Error(s) in loading state_dict` Error mentioned by @Samuel-gwb can be solved by including `strict=False` option in the `.load_state_dict()` function in the `query.py` file. However, I am not sure if this will affect the result of the protonation. Note that I only allow "cpu"  and block the "cuda" device to only allow cpu to help protonate the smi.

```python
class QueryModel:
    def __init__(self):

        self.models = []

        for i in range(25):
            model_name, model_class = "GINPair", GINPairV1
            model = model_class(
                num_node_features, num_edge_features, hidden_channels=96
            )
            base_path = path.dirname(__file__)
            # if torch.cuda.is_available() == False:  # If only CPU is available
            checkpoint = torch.load(
                    f"{base_path}/trained_model_without_epik/best_model_{i}.pt",
                    map_location=torch.device("cpu"),
                )
            # else:
            #     checkpoint = torch.load(
            #         f"{base_path}/trained_model_without_epik/best_model_{i}.pt"
            #     )

            model.load_state_dict(checkpoint["model_state_dict"], strict=False)
            model.eval()
            model.to(device=DEVICE)
            self.models.append(model)
```
I have found that using `Pool` for pkasolver can cause a multiprocessing error with the CUDA, since pkasolver calls for GPU and CPU with the `torch.device()` . The error is the `RuntimeError: Cannot re-initialize CUDA in forked subprocess. To use CUDA with multiprocessing, you must use the 'spawn' start method`. This can be solved by using `torch.multiprocessing.set_start_method()` to use the `spawn` method.

```python
from torch import multiprocessing as mp
def protonate_pkasolver(input_fname: str, output_fname: str, ncpu: int = 1):
    mp.set_start_method('spawn', force=True)
    from pkasolver.query import QueryModel
    model = QueryModel()
    pool = mp.Pool(ncpu)
    with open(output_fname, 'wt') as f:
        for smi, name in pool.imap_unordered(__protonate_pkasolver,
                                             ((mol, mol_name, model) for (mol, mol_name) in read_input(input_fname))):
            f.write(f'{smi}\t{name}\n')
```

However, there are other errors called after calling `torch.multiprocessing.Pool` which may be due to memory allocation -> 
`[W CudaIPCTypes.cpp:92] Producer process tried to deallocate over 1000 memory blocks referred by consumer processes. Deallocation might be significantly slowed down. We assume it will never going to be the case, but if it is, please file but to https://github.com/pytorch/pytorch
[W CudaIPCTypes.cpp:15] Producer process has been terminated before all shared CUDA tensors released. See Note [Sharing CUDA tensors]`

I am not sure how to deal with this.

Another funny thing is that the `dimorphite_dl` inside the `pkasolver` does not seem to store the output of the protonated_smiles. When running `run_dock`, it gives the following output

```
CCCCC
<rdkit.Chem.rdchem.Mol object at 0x7fab1e8c6a50>

For help, use: python dimorphite_dl.py --help

If you use Dimorphite-DL in your research, please cite:
Ropp PJ, Kaminsky JC, Yablonski S, Durrant JD (2019) Dimorphite-DL: An
open-source program for enumerating the ionization states of drug-like small
molecules. J Cheminform 11:14. doi:10.1186/s13321-019-0336-9.

CCCCCCC
<rdkit.Chem.rdchem.Mol object at 0x7f6d1d82fa50>

For help, use: python dimorphite_dl.py --help

If you use Dimorphite-DL in your research, please cite:
Ropp PJ, Kaminsky JC, Yablonski S, Durrant JD (2019) Dimorphite-DL: An
open-source program for enumerating the ionization states of drug-like small
molecules. J Cheminform 11:14. doi:10.1186/s13321-019-0336-9.
```

We may need to modify and fix these errors if we still want to pursue pkasolver as our alternative

**Comment by DrrDom** [comment]
Finally, I found an old environment with installed pkasolver.
Since as a backend for enumeration of protonation states it uses dimorphite (and includes it within the package as is), there is the same issue with command line arguments, when dimorphite catches all input arguments from the main program. So, `pkasolver` package should be modified, either to use the external custom `dimorphite` package or by fixing this issue inside the `pkasolver` package. I'll try to use the first option and modify the `pkasolver` package. This will result in a custom pkasolver version, which should be installed from my repo.

That is strange that this error did not occur for you.

I do not have GPU on my laptop, so I cannot test how it works with GPU. Before, I did not have errors due to multiprocessing, but maybe just because i do not have GPU.

I agree that switching to CPU mode for `pkasolver` is a good decision. We may customize further `pkasolver` package to enable this control for a user. For example, by adding an argument to `QueryModel.__init__`

**Comment by DrrDom** [comment]
OK, I got it why I had different issues, I previously modified `dimorphite`within `pkasolver` to solve the issue below.

Originally, `pkasolver` runs `dimorphite` via `subprocess`. This implementation is not compatible with `multiprocessing`, because every molecules is processed separately and the result is saved in the file with the same name `test.pkl` for all molecules. Therefore, it can be overwritten at any time or corrupted. I tried to use Python interface of dimorphite, but there is an issue with command lines arguments, which I did not observe previously for some reason.

Trying to import the necessary function from custom `Feriolet/dimorphite` package does not work as well. There is an argument `return_as_list` which is not available in command line args but used by the authors of dimorphite.

Finally:
- I made PR to `Feriolet/dimorphite` to fix parsing command line arguments. This is an ugly solution, but works for me.
- I pushed changes to `DrrDom/pkasolver` branch `noprints`, where I replaced built-in dimorphite with `Feriolet/dimophite`
- I pushed changes to `easydock` branch `pkasolver2`, where I disabled interception of strerr/stdout (there was an issue with the previous solution which I could not solve quickly)

Current installation:
pip install git+https://github.com/Feriolet/dimorphite_dl.git (after PR will be accepted)
pip install git+https://github.com/DrrDom/pkasolver.git@noprints
pip install git+https://github.com/ci-lab-cz/easydock.git@pkasolver2

Now it works for me. However, there are still errors in structures. 

Example:
input
`O=C(N1CCN(CC1)C(=O)C=2C=CC=C(C#CC3CC3)C2)C=4NN=C5CCCC45`
output
`O=C(c1cccc(C#CC2CC2)c1)N1CC[NH](C(=O)c2[nH]nc3c2CCC3)CC1`

It adds hydrogen to a nitrogen in a piperazine ring, but does not add charge. RDKit cannot parse this molecule as incorrect (a radical). Moreover, this protonation is definitely wrong, because this nitrogen cannot be protonated at pH 7.4. Such molecules are skipped from docking.

**Comment by DrrDom** [comment]
Please, check whether this implementation is compatible with computers with GPUs or there are still issues which should be solved.

**Comment by Feriolet** [comment]
I have not tested CPU, but the result from GPU seems to almost match with the smi that you tested. I have to add the `strict=False` in the QueryModel for it to work (not sure why it does not occur on your side).

I also have to modify some of the script to make the subprocessing works:

```python
def protonate_pkasolver(input_fname: str, output_fname: str, ncpu: int = 1):
    from pkasolver.query import QueryModel
    from torch import multiprocessing as mp
    mp.set_start_method('spawn', force=True)
    model = QueryModel()
    pool = mp.Pool(ncpu)
    with open(output_fname, 'wt') as f:
        for smi, name in pool.imap_unordered(partial(__protonate_pkasolver, model=model), read_input(input_fname)):
            f.write(f'{smi}\t{name}\n')
```
Note that I use the `mp.Pool(ncpu)` argument to access the GPU memory, which is not a good idea since it can lead to `out of memory` error if we use a high number of `ncpu`. If we still want to use GPU for this, we probably need to add a `ngpu` argument for this function (and its parent function).

Here is the result of the smi test:
![image](https://github.com/ci-lab-cz/easydock/assets/108213198/4805d5e1-6808-4bcc-8869-a2ed56ac900c)

I also tried to protonate 100 smiles used for my previous work (using `--protonation dimorphite`, and I notice that using `--protonation pkasovler` tends to deprotonate the smiles. I am not sure which one is correct. 

```
input: COC(=O)c1nc2[nH]ccc2cc1-c1cccc(C(C)C)c1
dimo : COC(=O)c1nc2[nH]ccc2cc1-c1cccc(C(C)C)c1
pkas : COC(=O)c1nc2[n-]ccc2cc1-c1cccc(C(C)C)c1

input: CC(C)c1cccc(-c2ncc(F)c3[nH]ccc23)c1
dimo : CC(C)c1cccc(-c2ncc(F)c3[nH]ccc23)c1
pkas : CC(C)c1cccc(-c2ncc(F)c3[n-]ccc23)c1
```

**Comment by Feriolet** [comment]
For CPU, I can reproduce the same result as the GPU. However, the CPU gives this error:
```Traceback (most recent call last):
  File "/home/user/miniforge3/envs/easydock/lib/python3.9/multiprocessing/pool.py", line 268, in __del__
  File "/home/user/miniforge3/envs/easydock/lib/python3.9/multiprocessing/queues.py", line 371, in put
AttributeError: 'NoneType' object has no attribute 'dumps'
```
I resolve this issue by using the `pool.close()` and `pool.join()` similar to the GPU ones.
I also need to use `strict=False` for this.

btw, some of the molecules print this while being protonated by the pkasolver. Is this something that we need to silent?
```
#########################
Could not identify any ionizable group. Aborting.
#########################
 ```

**Comment by DrrDom** [comment]
It seems that the calculation speed on CPU is sufficient. If so, we may force to use CPUs in `pkasolver` to avoid implementation issues with GPU. For example by adding `force_cpu` and call `model = QueryModel(force_cpu=True)`
```python
class QueryModel:
    def __init__(self, force_cpu=False):

        self.models = []

        for i in range(25):
            model_name, model_class = "GINPair", GINPairV1
            model = model_class(
                num_node_features, num_edge_features, hidden_channels=96
            )
            base_path = path.dirname(__file__)
            if not force_cpu and torch.cuda.is_available() == False:  # If only CPU is available
                checkpoint = torch.load(
                    f"{base_path}/trained_model_without_epik/best_model_{i}.pt",
                    map_location=torch.device("cpu"),
                )
            else:
                checkpoint = torch.load(
                    f"{base_path}/trained_model_without_epik/best_model_{i}.pt"
                )

            model.load_state_dict(checkpoint["model_state_dict"])
            model.eval()
            model.to(device=DEVICE)
            self.models.append(model)
```

I do not know why you need strict=False for CPU-mode. For me, both options work. So, we may add it if it necessary. However, i do not fully understand what issues it may cause in future.

Introducing `ngpu` argument will complicate the interface, because we will need to pass it through all function calls. And it will require a separate implementation. If this does not bring substantial speed advantage I would suggest to avoid this.

The errors caused by `multiprocessing` and fixing them with `pool.join()` and `pool.close()` are unexpected. Never met such issues. Below I attached my env configuration. Maybe you use more recent python or module version where some changes were made.

[easydock_env.txt](https://github.com/ci-lab-cz/easydock/files/14943120/easydock_env.txt)

The messages from pkasolver should be suppressed. I did it with `nostd` context, but now it raises an error and I did not have time to fix it. You may uncomment that line and test it. Or you may suggest another solution. As far as I remember, my solution intercepted only particular stdrerr/stdout, not globally.

I was more surprised with your output of protonated smiles. Below is my output for the same structures and I generally agree with that. There were a lot of differences. Did you use the latest easydock version from `noprints` branch?

![image](https://github.com/ci-lab-cz/easydock/assets/4629017/9d8e916d-dad9-4da9-b0e8-734cbc29da3a)

**Comment by Feriolet** [comment]
I finally can run it without any of those issue. I have reinstalled everything from scratch and the problem goes away. i guess there are conflicting torch version that I installed and messed up with the other packages.

For future reference:
```bash
conda create -n easydock -c conda-forge python=3.9 numpy=1.20 rdkit scipy dask distributed
conda activate easydock
pip install paramiko meeko vina
pip install git+https://github.com/Feriolet/dimorphite_dl.git
pip install git+https://github.com/DrrDom/pkasolver.git@noprints
pip install git+https://github.com/ci-lab-cz/easydock.git@pkasolver2
pip install torch==1.13.1+cpu  --extra-index-url https://download.pytorch.org/whl/cpu
pip install torch-geometric==2.0.1
pip install torch_scatter==2.1.1+pt113cpu -f https://data.pyg.org/whl/torch-1.13.1%2Bcpu.html
pip install torch_sparse==0.6.17+pt113cpu -f https://data.pyg.org/whl/torch-1.13.1%2Bcpu.html
pip install torch_spline_conv==1.2.2+pt113cpu -f https://data.pyg.org/whl/torch-1.13.1%2Bcpu.html
```

Yes, I agree that we should use CPU for convenience and consistency across other protonation software (dimorphite_dl and chemaxon). To silent the protonation output, I have added the contextlib package to the `add_protonation()` function inside the `database.py` file:

```python
                if program == 'chemaxon':
                    protonate_func = partial(protonate_chemaxon, tautomerize=tautomerize)
                    read_func = read_protonate_chemaxon
                elif program == 'dimorphite':
                    protonate_func = partial(protonate_dimorphite, ncpu=ncpu)
                    read_func = read_smiles
                elif program == 'pkasolver':
                    protonate_func = partial(protonate_pkasolver, ncpu=ncpu)
                    read_func = read_smiles
                else:
                    protonate_func = empty_func
                    read_func = empty_generator
                
                with contextlib.redirect_stdout(None):
                    protonate_func(input_fname=tmp.name, output_fname=output)
```

**Comment by DrrDom** [comment]
Many thanks for the installation notes. We will include them to the README.

Is now the output identical to mine, no difference in protonation states? If so, I will merge everything to master branches and add this solution with `contextlib`.

Finally, I will keep dimorphite implementation inside the code, but will remove it from the command line interface, because currently it is not useful and will confuse users only.

**Comment by Feriolet** [comment]
Yes the protonated smiles are identical to the most recent one you showed

**Comment by DrrDom** [comment]
One more question. Does it work on computers with GPU? Should we add `force_cpu` option or not?

**Comment by Feriolet** [comment]
I have not tested the GPU from scratch. It should also work given that it has the same result as CPU for the previous protonation (as in yesterday's result). I'll update you once I can test it on the GPU.

Assuming that the users will follow the torch installation, there may be no need to yse `force_cpu=True`. I guess it can be a good option if you want to make sure that people who accidentally installed torch-cuda got the warning to only use CPU.

**Comment by DrrDom** [comment]
I updated `easydock/pkasolver2` and `pkasolver/main`.

The minor issue which is remained - enumeration of stereoisomers after protonation, e.g. a new unspecified chiral center will appear in `C[C@@H]1CCCN(C)C1` after protonation. I'm thinking how to do that with minimal code perturbation and maximum flexibility for the future changes. It may worth to redesign `init_db` and pull the function `get_isomers` out of it and apply it only after protonated molecules were generated.

```python
if not os.path.isfile(args.output):
    create_db(args.output, args)
    init_db(args.output, args.input, args.prefix)
else:
    args_dict, tmpfiles = restore_setup_from_db(args.output)
    # this will ignore stored values of those args which were supplied via command line
    # command line args have precedence over stored ones
    for arg in supplied_args:
        del args_dict[arg]
    args.__dict__.update(args_dict)

dask_client = create_dask_client(args.hostfile)

if args.protonation:
    add_protonation(args.output, program=args.protonation, tautomerize=not args.no_tautomerization, ncpu=args.ncpu)

populate_stereoisomers(args.output, args.max_stereoisomers)
```

However, this will create an issue, that we will have records with identical `smi`, different `stereo_id` and different `protonated_smi`, that is very misleading and may result in many issues in future. a solution may be to introduce an additional filed to DB `protonated_id` and enable that a single molecule (smiles) may have several protonation states (which were not alternative protonation states in sense of dimorphite, but different stereoisomers appearing after protonation). I'm not confident with this solution, because it will complicate logic of functions and data manipulation. However, i do not see a better alternative.

Currently I tend to ignore this issue and postpone its solution for future.

**Comment by Samuel-gwb** [comment]
Some error use newst environment and run_dock ...  --protonation pkasolver ... :
RuntimeError: unable to open shared memory object </torch_4154596_2592645159_498> in read-write mode: Too many open files (24)

But it's successful to use:  run_dock ...  --protonation dimorphite ...

**Comment by Feriolet** [comment]
Have you tried reinstalling the conda environment from scratch? The error seems to be caused by torch.multiprocessing, but I am not sure if the default multiprocessing can call the torch multiprocessing.

Also, how many CPU did you use?

**Comment by Samuel-gwb** [comment]
Yes, I freshly installed a new conda environment, named as easydock_new.
My computer has 64 cpu. Need to specify cpu, something like cpu:0?

**Comment by Feriolet** [comment]
@DrrDom btw for your previous qn on GPU (if you are still interested):

Both GPU and CPU gives identical protonated smiles.
For 100 smiles:
`-c 30` protonates for 57.70 s
1 gpu protonates for 46.74s
2 gpu pool (shared gpu) protonates for 27.77s
4 gpu pool (shared gpu) protonates for 18.94s

**Comment by Samuel-gwb** [comment]
Command is as:
  run_dock -i "$smi_file" -o "$output_file" --program vina --config config_vina.yml --protonati
on pkasolver -c 1 --sdf

"--protonation dimorphite " is using the same input file. 

For QueryModel(), I think it should be created by "pip install ..." into the miniconda3/env/easydock_pka/lib/python3.9/site-packages/pkasolver/query.py. 
I have not modified it, which is as :
class QueryModel:
    def __init__(self):

        self.models = []

        for i in range(25):
            model_name, model_class = "GINPair", GINPairV1
            model = model_class(
                num_node_features, num_edge_features, hidden_channels=96
            )
            base_path = path.dirname(__file__)
            if torch.cuda.is_available() == False:  # If only CPU is available
                checkpoint = torch.load(
                    f"{base_path}/trained_model_without_epik/best_model_{i}.pt",
                    map_location=torch.device("cpu"),
                )
            else:
                checkpoint = torch.load(
                    f"{base_path}/trained_model_without_epik/best_model_{i}.pt"
                )

            model.load_state_dict(checkpoint["model_state_dict"])
            model.eval()
            model.to(device=DEVICE)
            self.models.append(model)

    def predict_pka_value(self, loader: DataLoader) -> np.ndarray:
        """
        ----------
        loader
            data to be predicted
        Returns
        -------
        np.array
            list of predicted pKa values
        """

        results = []
        assert len(loader) == 1
        for data in loader:  # Iterate in batches over the training dataset.
            data.to(device=DEVICE)
            consensus_r = []
            for model in self.models:
                y_pred = (
                    model(
                        x_p=data.x_p,
                        x_d=data.x_d,
                        edge_attr_p=data.edge_attr_p,
                        edge_attr_d=data.edge_attr_d,
                        data=data,
                    )
                    .reshape(-1)
                    .detach()
                )

                consensus_r.append(y_pred.tolist())
            results.extend(
                (
                    float(np.average(consensus_r, axis=0)),
                    float(np.std(consensus_r, axis=0)),
                )
            )
        return results







environment is as:
[easydock_pka.txt](https://github.com/ci-lab-cz/easydock/files/14953727/easydock_pka.txt)

**Comment by Feriolet** [comment]
I am assuming `easydock_pka` is the same as `easydock_new` environment?

**Comment by Samuel-gwb** [comment]
input files also attached :
[test.zip](https://github.com/ci-lab-cz/easydock/files/14954321/test.zip)


Error mesage:
################################
(easydock_pka) gwb@node01:~/RationalDesign/Small_Molecule/Y73C_GTP$ ./Ensemble_RunDock.sh
Traceback (most recent call last):
  File "/home/gwb/miniconda3/envs/easydock_pka/bin/run_dock", line 8, in <module>
    sys.exit(main())
  File "/home/gwb/miniconda3/envs/easydock_pka/lib/python3.9/site-packages/easydock/run_dock.py", line 207, in main
    add_protonation(args.output, program=args.protonation, tautomerize=not args.no_tautomerization, ncpu=args.ncpu)
  File "/home/gwb/miniconda3/envs/easydock_pka/lib/python3.9/site-packages/easydock/database.py", line 348, in add_protonation
    protonate_func(input_fname=tmp.name, output_fname=output)
  File "/home/gwb/miniconda3/envs/easydock_pka/lib/python3.9/site-packages/easydock/protonation.py", line 92, in protonate_pkasolver
    for smi, name in pool.imap_unordered(partial(__protonate_pkasolver, model=model), read_input(input_fname)):
  File "/home/gwb/miniconda3/envs/easydock_pka/lib/python3.9/multiprocessing/pool.py", line 870, in next
    raise value
  File "/home/gwb/miniconda3/envs/easydock_pka/lib/python3.9/multiprocessing/pool.py", line 537, in _handle_tasks
    put(task)
  File "/home/gwb/miniconda3/envs/easydock_pka/lib/python3.9/multiprocessing/connection.py", line 206, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/home/gwb/miniconda3/envs/easydock_pka/lib/python3.9/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
  File "/home/gwb/miniconda3/envs/easydock_pka/lib/python3.9/site-packages/torch/multiprocessing/reductions.py", line 367, in reduce_storage
    df = multiprocessing.reduction.DupFd(fd)
  File "/home/gwb/miniconda3/envs/easydock_pka/lib/python3.9/multiprocessing/reduction.py", line 198, in DupFd
    return resource_sharer.DupFd(fd)
  File "/home/gwb/miniconda3/envs/easydock_pka/lib/python3.9/multiprocessing/resource_sharer.py", line 48, in __init__
    new_fd = os.dup(fd)
OSError: [Errno 24] Too many open files
(easydock_pka) gwb@node01:~/RationalDesign/Small_Molecule/Y73C_GTP$

################################

**Comment by Samuel-gwb** [comment]
> I am assuming `easydock_pka` is the same as `easydock_new` environment?

Yes, they are the same. typo mistakes

**Comment by Feriolet** [comment]
Yes, it still runs without issue.

Ok, what about changing the protonation function. Maybe it works for your case?

```python
def protonate_pkasolver(input_fname: str, output_fname: str, ncpu: int = 1):
    from pkasolver.query import QueryModel
    model = QueryModel()
    with contextlib.redirect_stdout(None):
        pool = mp.Pool(ncpu)
        with open(output_fname, 'wt') as f:
            pkasolver_output = pool.imap_unordered(partial(__protonate_pkasolver, model=model), read_input(input_fname))
            pool.close()
            pool.join()
            for smi, name in pkasolver_output:
                f.write(f'{smi}\t{name}\n')
```

**Comment by DrrDom** [comment]
@Samuel-gwb, I'm a little bit lost. You posted two error messages with "too many open files". One is related to `torch`, another to standard `multiprocessing`. Do you use the latest version `easydock/pkasolver2` branch? Do you have GPU?

Since you use `-c 1` I cannot imagine how you may exceed the number of opened files.

You may increase the number of file descriptors opened simultaneously `ulimit -n 4096`, but this looks like not a proper solution.

**Comment by Feriolet** [comment]
Yes that is what I thought as well.

From the error, it looks like the default multiprocessing calls to torch version, which calls the default version again.
It is very interesting.

I tried not to use the ulimit solution as it is surprising that accessing one cpu would cause this issue and may be used as a last resort if everything else fails

**Comment by DrrDom** [comment]
If `multiprocessing.pool` calls `multiprocessing.pool` directly it should result in an error about nested processes or the like, because this is forbidden by design. If this call happens through `torch`, maybe this avoids this error, but causes another one.

In that case I see two possible solutions:
1. Add `force_cpu` argument to `QueryModel` and set it to `True`.
2. Detect GPUs within `protonate_pkasolver` function and call protonation without `multiprocessing.pool`

@Samuel-gwb, could you test the function below?

```python
def protonate_pkasolver(input_fname: str, output_fname: str, ncpu: int = 1):
    import torch
    from pkasolver.query import QueryModel
    model = QueryModel()
    with contextlib.redirect_stdout(None):
        if torch.cuda.is_available() or ncpu == 1:
            with open(output_fname, 'wt') as f:
                for mol, mol_name in read_input(input_fname):
                    smi, name = _protonate_pkasolver(mol, mol_name, model=model)
                    f.write(f'{smi}\t{name}\n')```            
        else:
            pool = Pool(ncpu)
            with open(output_fname, 'wt') as f:
                for smi, name in pool.imap_unordered(partial(__protonate_pkasolver, model=model), read_input(input_fname)):
                    f.write(f'{smi}\t{name}\n')```

**Comment by Samuel-gwb** [comment]
Yes, I use the same easydock_pka environment for different tests. And the last error message can be repeated for last several times.
Will  try your solutions with modified pka_solver function!

**Comment by Samuel-gwb** [comment]
Very confused !

Again, freshly installed an environment, just change easydock --> easydock_test1 :
--------------------------------------------------
conda create -n easydock_test1 -c conda-forge python=3.9 numpy=1.20 rdkit scipy dask distributed
conda activate easydock_test1
pip install paramiko meeko vina
pip install git+https://github.com/Feriolet/dimorphite_dl.git
pip install git+https://github.com/DrrDom/pkasolver.git@noprints
pip install git+https://github.com/ci-lab-cz/easydock.git@pkasolver2
pip install torch==1.13.1+cpu  --extra-index-url https://download.pytorch.org/whl/cpu
pip install torch-geometric==2.0.1
pip install torch_scatter==2.1.1+pt113cpu -f https://data.pyg.org/whl/torch-1.13.1%2Bcpu.html
pip install torch_sparse==0.6.17+pt113cpu -f https://data.pyg.org/whl/torch-1.13.1%2Bcpu.html
pip install torch_spline_conv==1.2.2+pt113cpu -f https://data.pyg.org/whl/torch-1.13.1%2Bcpu.html
pip install molvs chembl_webresource_client matplotlib pytest-cov codecov svgutils cairosvg ipython
--------------------------------------------------

Then,Use default protonate_pkasolver function, error:
--------------------------------------------------
Traceback (most recent call last):
  File "/home/gwb/miniconda3/envs/easydock_test1/bin/run_dock", line 8, in <module>
    sys.exit(main())
  File "/home/gwb/miniconda3/envs/easydock_test1/lib/python3.9/site-packages/easydock/run_dock.py", line 207, in main
    add_protonation(args.output, program=args.protonation, tautomerize=not args.no_tautomerization, ncpu=args.ncpu)
  File "/home/gwb/miniconda3/envs/easydock_test1/lib/python3.9/site-packages/easydock/database.py", line 348, in add_protonation
    protonate_func(input_fname=tmp.name, output_fname=output)
  File "/home/gwb/miniconda3/envs/easydock_test1/lib/python3.9/site-packages/easydock/protonation.py", line 92, in protonate_pkasolver
    for smi, name in pool.imap_unordered(partial(__protonate_pkasolver, model=model), read_input(input_fname)):
  File "/home/gwb/miniconda3/envs/easydock_test1/lib/python3.9/multiprocessing/pool.py", line 870, in next
    raise value
  File "/home/gwb/miniconda3/envs/easydock_test1/lib/python3.9/multiprocessing/pool.py", line 537, in _handle_tasks
    put(task)
  File "/home/gwb/miniconda3/envs/easydock_test1/lib/python3.9/multiprocessing/connection.py", line 206, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/home/gwb/miniconda3/envs/easydock_test1/lib/python3.9/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
  File "/home/gwb/miniconda3/envs/easydock_test1/lib/python3.9/site-packages/torch/multiprocessing/reductions.py", line 367, in reduce_storage
    df = multiprocessing.reduction.DupFd(fd)
  File "/home/gwb/miniconda3/envs/easydock_test1/lib/python3.9/multiprocessing/reduction.py", line 198, in DupFd
    return resource_sharer.DupFd(fd)
  File "/home/gwb/miniconda3/envs/easydock_test1/lib/python3.9/multiprocessing/resource_sharer.py", line 48, in __init__
    new_fd = os.dup(fd)
OSError: [Errno 24] Too many open files
--------------------------------------------------


Use Feriolet‘s version to modify "/home/gwb/miniconda3/envs/easydock_test1/lib/python3.9/site-packages/easydock/protonation.py", replace contents of 'def protonate_pkasolver', then:
--------------------------------------------------
Traceback (most recent call last):
  File "/home/gwb/miniconda3/envs/easydock_test1/bin/run_dock", line 8, in <module>
    sys.exit(main())
  File "/home/gwb/miniconda3/envs/easydock_test1/lib/python3.9/site-packages/easydock/run_dock.py", line 207, in main
    add_protonation(args.output, program=args.protonation, tautomerize=not args.no_tautomerization, ncpu=args.ncpu)
  File "/home/gwb/miniconda3/envs/easydock_test1/lib/python3.9/site-packages/easydock/database.py", line 348, in add_protonation
    protonate_func(input_fname=tmp.name, output_fname=output)
  File "/home/gwb/miniconda3/envs/easydock_test1/lib/python3.9/site-packages/easydock/protonation.py", line 106, in protonate_pkasolver
    for smi, name in pkasolver_output:
  File "/home/gwb/miniconda3/envs/easydock_test1/lib/python3.9/multiprocessing/pool.py", line 870, in next
    raise value
  File "/home/gwb/miniconda3/envs/easydock_test1/lib/python3.9/multiprocessing/pool.py", line 537, in _handle_tasks
    put(task)
  File "/home/gwb/miniconda3/envs/easydock_test1/lib/python3.9/multiprocessing/connection.py", line 206, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/home/gwb/miniconda3/envs/easydock_test1/lib/python3.9/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
  File "/home/gwb/miniconda3/envs/easydock_test1/lib/python3.9/site-packages/torch/multiprocessing/reductions.py", line 366, in reduce_storage
    fd, size = storage._share_fd_cpu_()
RuntimeError: unable to open shared memory object </torch_55536_471936146_501> in read-write mode: Too many open files (24)
--------------------------------------------------


Use Pavel's version, need to modify 'smi, name = _protonate_...' --> 'smi, name = __protonate_...' :
--------------------------------------------------
Traceback (most recent call last):
  File "/home/gwb/miniconda3/envs/easydock_test1/bin/run_dock", line 8, in <module>
    sys.exit(main())
  File "/home/gwb/miniconda3/envs/easydock_test1/lib/python3.9/site-packages/easydock/run_dock.py", line 207, in main
    add_protonation(args.output, program=args.protonation, tautomerize=not args.no_tautomerization, ncpu=args.ncpu)
  File "/home/gwb/miniconda3/envs/easydock_test1/lib/python3.9/site-packages/easydock/database.py", line 348, in add_protonation
    protonate_func(input_fname=tmp.name, output_fname=output)
  File "/home/gwb/miniconda3/envs/easydock_test1/lib/python3.9/site-packages/easydock/protonation.py", line 119, in protonate_pkasolver
    smi, name = __protonate_pkasolver(mol, mol_name, model=model)
TypeError: __protonate_pkasolver() got multiple values for argument 'model'
--------------------------------------------------

**Comment by Feriolet** [comment]
Please put a bracket for the mol and mol_name

```python
smi, name = __protonate_pkasolver((mol, mol_name), model=model)
```

**Comment by Samuel-gwb** [comment]
Great, it works! Many thanks !


1) One more thing,  -c have to be 1. 
run_dock -i GTP.smi -o GTP_vina.db --program vina --config config_vina.yml --protonation pkasolver -c 1 --sdf

Any -c > 1 will cause "Too many open files" error.

2) Another thing is that, additional nitrogen adjacent to the imidazole ring of GTP was protonated as NH-. 
I know that it will be NH2 when someone use schrodinger.

GTP smi:
OC1C(COP(=O)(OP(=O)(OP(=O)(O)O)O)O)OC(C1O)n1cnc2c1[nH]c(N)nc2=O
protonated by pkasolver -->
[NH-]c1nc(=O)c2ncn([C@H]3O[C@@H](CO[P@](=O)([O-])O[P@](=O)([O-])OP(=O)([O-])[O-])[C@H](O)[C@H]3[O-])c2[n-]1

[GTP_vina.sdf.txt](https://github.com/ci-lab-cz/easydock/files/14964560/GTP_vina.sdf.txt)

**Comment by Feriolet** [comment]
1. Yes, we have made it that `multiprocessing` is used when `-c` > 1, which is the package that is giving you the error. 

We hope that using 1 cpu is sufficient for your use case. I honestly still can't reproduce your error, so I can't really help you too much with that. I also tried running it on Apple M1, and there is also no such issue. We can still try to tackle the `multiprocessing` issue if you wish to access more than 1 cpu, but it may be challenging as some of the obvious solutions do not work.

2. Regarding this, easydock only give 1 protonated smiles. I am sure GTP has multiple protonation center, and it just happens the pkasolver give the [NH-] out of the many possible protonated smiles. We are unsure if we should give multiple protonated smiles per smile input, as we are still considering the stereoisomer enumeration issue after protonation.

@DrrDom correct me if there is any mistake I said

**Comment by DrrDom** [comment]
The error for ncpu > 1 is strange. This means that you do not have detectable GPU and use exclusively `multiprocessing`. I never met such en error for multiprocessing.

Wrong protonations may occur. Every protonation tool is incorrect to some extent. `pkasolver` model publicly available was trained on single-center molecules. Therefore, prediction for complex molecules with multiple protonation centers may be incorrect. That is why applicability of different protonation tools should be studied more thoroughly. Meanwhile we may use `pkasolver` as alternative to chemaxon.

**Comment by DrrDom** [comment]
I updated `master` with the most recent changes. I'll keep the issue open, because I believe we will return to it in future.
Thanks a lot to everybody who helped with that!

**Comment by Samuel-gwb** [comment]
Great !
Some tiny things:
1) in readme, the line for pip installaltion of torch_spline_conv contains additional ''' at the end.

2) It seems that "pip install cairosvg svgutils" is needed. 
    And, at last, installation may need include "pip install ." at $easydock_home.

3) Thus, when using cpu-based pkasolver for protonation, one need set "-c 1" ? If so, include it in readme?

**Comment by DrrDom** [comment]
Thank you!
1. Fixed. 
2. Installation of `easydock` was described. Your suggestion is not relevant for ordinary users. This is mainly for developers, who has a clone of the repository.
3. Currently it is not necessary. Your case seems very specific. We will collect other user responses whether they have issues with that. However, it may worth to mention this issue in README to attract user attention.

**Comment by Feriolet** [comment]
I agreed with @Samuel-gwb for his 2nd point. Don't we need the `pip install cairosvg svgutils` to run the pkasolver? At least for my side it gave the import error for the cairosvg

**Comment by DrrDom** [comment]
You were right) Thanks to pointing me out. I indeed missed to add these packages `cairosvg svgutils` to the list of required ones. I'll do that.

**Comment by DrrDom** [comment]
The issue with protonation was solved to some extent by integration of molgpka. Molgpka model was trained on chemaxon predictions and generally aligns well with chemaxon. There are some systematic errors. Some of them were solved by post-processing, some cannot be solved. Sometimes aliphatic nitrogens are not protonated and sometimes they are excessively protonated (almost every nitrogen). Phenols are also a problem, many of them are deprotonated, however, there are many borderline cases. The version with molgpka is in `dev` branch and I expect soon we will release the next version, which has many changes.

**Comment by DrrDom** [comment]
MolGpKa was integrated into the release version. Protonation states differs from chemaxon for ~40% of molecules which have ionogenic centers. There are some errors of molgpka (more often) or chemaxon (less often). Also the difference was caused by protonation of alternative centers in molecules with several possible ones, e.g. non-symmetricaly substituted piperizines. We implemented some fixes to avoid obvious errors of molgpka, but this improved states for ~5% cases.

I will not close the topic, because if someone will recommend other open-source tools which are more stable and consistent, we may consider their integration.

**Comment by DrrDom** [comment]
Just to be fully precise. Those SMILES were not from Chemaxon, but from MolGpKa + correction rules. I added more explicit explanation to the previous message.

It seems that pka-predictor has issues with guanidine and amidine groups. Sometimes it protonates a nitrogen in anilines, while there are other more basic nitrogens in a structure. The same happens for amide groups which may be protonated while there is a secondary amino group in the structure, which should be protonated instead. However, it does not protonate aromatic nitrogens. Polyphenols are always tricky to predict.

I'll make and share a data set of structures from ChEMBL and predicted protonation forms by tools we tested.

**Comment by DrrDom** [comment]
Here is a dataset, which we used for testing of different approaches - https://doi.org/10.5281/zenodo.17179369. There are also outputs for some programs. As new results will become available, we can update the files.

**Comment by Feriolet** [comment]
Oh right my bad. Yes, they did modify the smiles by removing the explicit hydrogens. But since I am running through multiprocessing, it is hard to backtrack which ID it is until I finished rerunning the script.

```python
    # Removing explicit hydrogens:
    # Removing the hydrogen bonds (eg: [OH:15] should become O
    for j in range(len(smiles)):
        if j == len(smiles):
            break

        if smiles[j] == ':':
            pos = 0
            for k in range(j + 1, len(smiles)):
                if isDigit(smiles[k]) is False:
                    pos = k
                    break
            smiles = smiles[:j] + smiles[pos:]
            smiles = smiles.replace('[N]', 'N').replace('[n]', 'n').replace('[O]', 'O')

    # Removing explicit hydrogens:
    smiles = smiles.replace('([H])', '')
    smiles = smiles.replace('([H])', '')
    smiles = smiles.replace('[C-]', 'C')
    smiles = smiles.replace('[n]', 'n')
    # Add explicit hydrogens for terminal nitrogens
    smiles = smiles.replace('(N)', '([NH2])')

    # Checking the validity of the SMILES string
    mol_test = Chem.MolFromSmiles(smiles, sanitize=True)

    if mol_test is None and smiles != '':
        print('| Invalid SMILES: %106s |' % smiles)
        raise TypeError

    # Loading the molecule to check some properties
    mol_original = Chem.MolFromSmiles(smiles, sanitize=False)
    Chem.rdmolops.RemoveHs(mol_original, sanitize=False)
```

**Comment by Feriolet** [comment]
Yes, I also found it weird why they decided to remove the explicit hydrogens. I tried reading the supplementary docs also, but they did not mention anything about explocit hydrogens. When I leave my laptop running for 7 hours, I estimate that there are over 50 smiles with incorrect parsing, which means something wrong with the code removing the hydrogens.

Hence, I will try to see if they did the same encoding for their experimental smiles used for training as well. If they do, then it may be difficult to fix it without retraining the model. 


P.S. We reached 100 comments for this issue thread :)

**Comment by DrrDom** [comment]
If the error is rare we can analyze outputs without fixing this issue. If protonation states will be reasonable and better than for molgpka, then the toll can be considered useful and we may integrate the current version. However, since it seems it is in a very early stage of development, we may need to fix the version and approve new versions after tests.

So, there is no need to fix the error right now.

**Comment by DrrDom** [comment]
Thank you!
I re-generate all SMILES to fix the same RDKit version 2025.03.5 and will update files on zenodo asap. Below is statistics for structures which were changed in at least one program (895229 out of 1748091). So far, molgpka with fixes implemented in easydock gives the best correspondence with chemaxon, the column `n_changed&chemaxon`.
```
 name               n_total n_unchanged  n_unchanged&chemaxon  n_changed   n_changed&chemaxon   n_changed&!chemaxon
 molgpka             895229      374024                 283121    521205               375952                145253
 molgpka_easydock    895229      379084                 286204    516145               399227                116918
 pka-predictor       895229      199025                 116268    696204               359991                336213
 pkasolver_easydock  895229      303463                 199076    591766               317896                273870
```
pka-predictor gives the lowest correspondence. While there are some borderline cases, there are many true error. An example is CHEMBL100022 - it ignores guanidino group but protonates amide.

<img width="846" height="142" alt="Image" src="https://github.com/user-attachments/assets/3d66004f-ef6f-4446-8bfd-f7d9b933df1c" />
